src/Features/VisualBasic/Portable/ConvertForToForEach/VisualBasicConvertForToForEachCodeRefactoringProvider.vb

```diff
@@ -0,0 +1,102 @@
+﻿' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+Imports System.Composition
+Imports System.Threading
+Imports Microsoft.CodeAnalysis.CodeRefactorings
+Imports Microsoft.CodeAnalysis.ConvertForToForEach
+Imports Microsoft.CodeAnalysis.Options
+Imports Microsoft.CodeAnalysis.Text
+Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
+
+Namespace Microsoft.CodeAnalysis.VisualBasic.ConvertForToForEach
+    <ExportCodeRefactoringProvider(LanguageNames.VisualBasic, Name:=NameOf(VisualBasicConvertForToForEachCodeRefactoringProvider)), [Shared]>
+    Friend Class VisualBasicConvertForToForEachCodeRefactoringProvider
+        Inherits AbstractConvertForToForEachCodeRefactoringProvider(Of
+            StatementSyntax,
+            ForBlockSyntax,
+            ExpressionSyntax,
+            MemberAccessExpressionSyntax,
+            TypeSyntax,
+            VariableDeclaratorSyntax)
+
+        Protected Overrides Function GetTitle() As String
+            Return VBFeaturesResources.Convert_For_to_For_Each
+        End Function
+
+        Protected Overrides Function IsValidCursorPosition(forBlock As ForBlockSyntax, cursorPos As Integer) As Boolean
+            ' If there isn't a selection, then we allow the refactoring in the 'for' keyword, or at the end
+            ' of hte for-statement signature.
+
+            Dim forStatement = forBlock.ForStatement
+            Dim startSpan = forStatement.ForKeyword.Span
+            Dim endSpan = TextSpan.FromBounds(forStatement.Span.End, forStatement.FullSpan.End)
+
+            Return startSpan.IntersectsWith(cursorPos) OrElse endSpan.IntersectsWith(cursorPos)
+        End Function
+
+        Protected Overrides Function GetBodyStatements(forStatement As ForBlockSyntax) As SyntaxList(Of StatementSyntax)
+            Return forStatement.Statements
+        End Function
+
+        Protected Overrides Function IsValidVariableDeclarator(firstVariable As VariableDeclaratorSyntax) As Boolean
+            ' we only support local declarations in vb that have a single identifier.  i.e.
+            '    dim x = list(i)
+            Return firstVariable.Names.Count = 1
+        End Function
+
+        Protected Overrides Function TryGetForStatementComponents(
+                forBlock As ForBlockSyntax,
+                ByRef iterationVariable As SyntaxToken,
+                ByRef initializer As ExpressionSyntax,
+                ByRef memberAccess As MemberAccessExpressionSyntax,
+                ByRef stepValueExpressionOpt As ExpressionSyntax,
+                cancellationToken As CancellationToken) As Boolean
+
+            Dim forStatement As ForStatementSyntax = forBlock.ForStatement
```

Comment:
so thing like  parenthese will block it from offering refactoring? I am fine, just asking to see whether using IOperation can improve over syntax.

Reply:
Yes.  We look for specific patterns.  However, IOperation would not help, as we would still have to drill into IParenthesizedOperation.  So, in both cases, we have to decide if the parenthesized forms are things we want to accept.